### PR TITLE
Quaternions

### PR DESCRIPTION
**Pull request by abisner**
This introduces a new `Quaternion` type to `Picasso::BatchedLinearAlgebra`. The quaternions comprise a special non-commutative algebra useful for efficiently representing rotation matrices as a 4-vector:

$`q=(w,x,y,z)`$

or in axis-angle notation:

$`q=(\phi,\pmb{a})`$
A where $`\phi`$, $`\pmb{a}`$ are the angle and axis of rotation, respectively.

In addition to representing rotations, they are useful for Jacobi iterations in an SVD decomposition algorithm.

### CODE REVIEW

File: unit_test/tstBatchedLinearAlgebra.hpp [hunk 1]
```diff
@@ -1134,6 +1134,153 @@ void vectorTest()
...
+    // Check deep copy.
+    LinearAlgebra::Quaternion<double> w;
+    LinearAlgebra::deepCopy( w, q );
+    for ( int i = 0; i < 4; ++i )
+        EXPECT_DOUBLE_EQ( w( i ), 32.3 );
+}
+
+void quaMatRotTest()
+{
+    constexpr double pi = Kokkos::Experimental::pi_v<double>;
```
Comment (streeve) [inline unit_test/tstBatchedLinearAlgebra.hpp:1241]: Would need to require 3.6 to use this

File: src/Picasso_FieldTypes.hpp [hunk 2]
```diff
@@ -203,7 +205,7 @@ struct Scalar : ScalarBase
     template <class U>
     using field_type = Scalar<U>;
     template <int NumSpaceDim>
-    using gradient_type = Vector<T, NumSpaceDim>;
+    using gradient_type = LinearAlgebra::Vector<T, NumSpaceDim>;
```
Comment (streeve) [inline src/Picasso_FieldTypes.hpp:208]: Shouldn't this be a separate changeset?

Comment (abisner) [reply src/Picasso_FieldTypes.hpp:208]: Yes, this one made it in on accident.

**Comment by streeve** [comment]
@stvdwtt maybe of interest to you..

**Comment by stvdwtt** [comment]
Very nice that you're adding support for quaternions!

**Review by streeve — COMMENTED** [review]
I think you mentioned you had one more easy test, but otherwise seems ready to go

File: src/Picasso_FieldTypes.hpp [hunk 3]
```diff
@@ -237,6 +239,21 @@ struct Vector : VectorBase
...
+    static constexpr int rank = 1;
+    static constexpr int size = 4;
+    static constexpr int dim0 = 4;
+    using data_type = value_type[4];
+    using linear_algebra_type = LinearAlgebra::QuaternionView<T>;
+    template <class U>
+    using field_type = Quaternion<U>;
+    template <class U>
+    using matrix_type = Mat3<U>;
+};
```
[the remark this answers is not in the document]
Comment (abisner) [reply src/Picasso_FieldTypes.hpp:255]: Not for a `Quaternion` type, no. I'm not sure what the generalization of gradients to complex numbers or quaternions looks like. And I don't have an immediate use case.

File: unit_test/tstBatchedLinearAlgebra.hpp [hunk 4]
```diff
@@ -1134,6 +1134,152 @@ void vectorTest()
...
+    EXPECT_EQ( x_view.stride_0(), 1 );
+    EXPECT_EQ( x_view.stride( 0 ), 1 );
+    EXPECT_EQ( x_view.extent( 0 ), 4 );
+
+    EXPECT_EQ( x_view( 0 ), 1.2 );
+    EXPECT_EQ( x_view( 1 ), -3.5 );
+    EXPECT_EQ( x_view( 2 ), 5.4 );
+    EXPECT_EQ( x_view( 3 ), -2.4 );
+
+    // Check a deep copy
```
[the remark this answers is not in the document]
Comment (abisner) [reply unit_test/tstBatchedLinearAlgebra.hpp:1162]: Yes, I think it was incorrect in the vector test comment as well, which I cut-and-paste.

**Comment by streeve** [comment]
Merge when you're ready @abisner

**Comment by streeve** [comment]
HIP timeout unrelated
